domain/relation/service/relation.go

```diff
@@ -962,3 +983,90 @@ func (s *Service) SetRelationUnitSettings(
 	}
 	return s.st.SetRelationUnitSettings(ctx, relationUnitUUID, settings)
 }
+
+// ImportRelations sets relations imported in migration. It first builds all the
+// relations to insert from the arguments, then inserts them at the end so as to
+// wait as long as possible before turning into a write transaction.
+func (s *Service) ImportRelations(ctx context.Context, args relation.ImportRelationsArgs) error {
+	for _, arg := range args {
+		relUUID, err := s.importRelation(ctx, arg)
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		for _, ep := range arg.Endpoints {
+			err = s.importRelationEndpoint(ctx, relUUID, ep)
+			if err != nil {
+				return errors.Capture(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Service) importRelation(ctx context.Context, arg relation.ImportRelationArg) (corerelation.UUID, error) {
+	var relUUID corerelation.UUID
+
+	eps := arg.Key.EndpointIdentifiers()
+
+	switch len(eps) {
+	case 1:
+		// Peer relations are implicitly imported during migration of applications
+		// during the call to CreateApplication.
+		var err error
+		relUUID, err = s.st.GetPeerRelationUUIDByEndpointIdentifiers(ctx, eps[0])
+		if err != nil {
+			return relUUID, errors.Errorf("getting peer relation %d by endpoint %q: %w", arg.ID, eps[0], err)
+		}
+	case 2:
+		idep1, err := relation.NewCandidateEndpointIdentifier(eps[0].String())
+		if err != nil {
+			return relUUID, errors.Errorf("parsing endpoint identifier %q: %w", eps[0].String(), err)
+		}
+		idep2, err := relation.NewCandidateEndpointIdentifier(eps[1].String())
+		if err != nil {
+			return relUUID, errors.Errorf("parsing endpoint identifier %q: %w", eps[1].String(), err)
+		}
+
+		relUUID, err = s.st.SetRelationWithID(ctx, idep1, idep2, uint64(arg.ID))
```

Comment:
Can `GetPeerRelationUUIDByEndpointIdentifiers`  go down into this one? Then we can do it in one transaction.

Reply:
I'd rather not as we're not setting a peer relation during this import.

Reply:
Ah, I think I misread the code, ignore that comment.